src/librustc_typeck/check/cast.rs

```diff
@@ -41,155 +106,270 @@ impl<'tcx> CastCheck<'tcx> {
+                                            fcx.infcx().ty_to_string(t_1)));
+        }
+
     }
 
-    let span = cast.span;
-    let e = &cast.expr;
-    let t_e = structurally_resolved_type(fcx, span, cast.expr_ty);
-    let t_1 = structurally_resolved_type(fcx, span, cast.cast_ty);
-    let tcx = fcx.tcx();
-
-    // Check for trivial casts.
-    if !ty::type_has_ty_infer(t_1) {
-        if let Ok(()) = coercion::mk_assignty(fcx, e, t_e, t_1) {
-            if ty::type_is_numeric(t_1) && ty::type_is_numeric(t_e) {
-                tcx.sess.add_lint(lint::builtin::TRIVIAL_NUMERIC_CASTS,
-                                  e.id,
-                                  span,
-                                  format!("trivial numeric cast: `{}` as `{}`. Cast can be \
-                                           replaced by coercion, this might require type \
-                                           ascription or a temporary variable",
-                                          fcx.infcx().ty_to_string(t_e),
-                                          fcx.infcx().ty_to_string(t_1)));
-            } else {
-                tcx.sess.add_lint(lint::builtin::TRIVIAL_CASTS,
-                                  e.id,
-                                  span,
-                                  format!("trivial cast: `{}` as `{}`. Cast can be \
-                                           replaced by coercion, this might require type \
-                                           ascription or a temporary variable",
-                                          fcx.infcx().ty_to_string(t_e),
-                                          fcx.infcx().ty_to_string(t_1)));
+    pub fn check<'a>(mut self, fcx: &FnCtxt<'a, 'tcx>) {
+        self.expr_ty = structurally_resolved_type(fcx, self.span, self.expr_ty);
+        self.cast_ty = structurally_resolved_type(fcx, self.span, self.cast_ty);
+
+        debug!("check_cast({}, {} as {})", self.expr.id, self.expr_ty.repr(fcx.tcx()),
+               self.cast_ty.repr(fcx.tcx()));
+
+        if ty::type_is_error(self.expr_ty) || ty::type_is_error(self.cast_ty) {
+            // No sense in giving duplicate error messages
+        } else if self.try_coercion_cast(fcx) {
+            self.trivial_cast_lint(fcx);
+            debug!(" -> CoercionCast");
+            fcx.tcx().cast_kinds.borrow_mut().insert(self.expr.id,
+                                                     CastKind::CoercionCast);
+        } else { match self.do_check(fcx) {
+            Ok(k) => {
+                debug!(" -> {:?}", k);
+                fcx.tcx().cast_kinds.borrow_mut().insert(self.expr.id, k);
             }
-            return;
+            Err(e) => self.report_cast_error(fcx, e)
+        };}
+    }
+
+    /// Check a cast, and report an error if one exists. In some cases,
+    /// this can return Ok and create type errors rather than returning
+    /// directly. coercion-cast is handled in check instead of here.
```

Comment:
It would be good to explain how this happens since there is no obvious errors added in the function and all its inputs are immutable.